### PR TITLE
HV-1155 Class loading issues in modular environments

### PR DESCRIPTION
**Pull request by gsmet**
* https://hibernate.atlassian.net/browse/HV-1155

Long story short...

### CODE REVIEW

File: osgi/integrationtest/src/test/java/org/hibernate/validator/osgi/integrationtest/OsgiIntegrationTest.java [hunk 1]
```diff
@@ -212,7 +215,7 @@ private ExpressionFactory buildExpressionFactory() {
 		Thread.currentThread().setContextClassLoader( getClass().getClassLoader() );
 
 		try {
-			return ExpressionFactory.newInstance();
+			return ELManager.getExpressionFactory();
```
Comment (gunnarmorling) [inline osgi/integrationtest/src/test/java/org/hibernate/validator/osgi/integrationtest/OsgiIntegrationTest.java]: Couldn't we even simplify that entire test by simply relying on the default interpolator?

Comment (gsmet) [inline osgi/integrationtest/src/test/java/org/hibernate/validator/osgi/integrationtest/OsgiIntegrationTest.java]: I supposed the purpose of this test was to check that passing an external `ExpressionFactory` would work.

File: engine/src/main/java/org/hibernate/validator/internal/util/privilegedactions/SetClassLoader.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import java.security.PrivilegedAction;
+
+import org.hibernate.validator.internal.util.Contracts;
+
+/**
+ * Privileged action used to set the Thread context class loader.
+ *
+ * @author Guillaume Smet
+ */
+public final class SetClassLoader implements PrivilegedAction<Void> {
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/util/privilegedactions/SetClassLoader.java]: I'd name it `SetContextClassLoader`.

File: engine/src/main/java/org/hibernate/validator/internal/util/privilegedactions/SetClassLoader.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+	private final ClassLoader classLoader;
+
+	public static SetClassLoader ofContext(ClassLoader classLoader) {
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/util/privilegedactions/SetClassLoader.java]: We named these methods `action` in the case where there are not several ones with different semantics.

File: osgi/integrationtest/src/test/java/org/hibernate/validator/osgi/integrationtest/OsgiIntegrationTest.java [hunk 4]
```diff
@@ -116,7 +119,7 @@ public void canObtainValidatorFactoryAndPerformValidation() {
 				.validate( new Customer() );
 
 		assertEquals( 1, constraintViolations.size() );
-		assertEquals( "must be greater than or equal to 1", constraintViolations.iterator().next().getMessage() );
+		assertEquals( "must be greater than or equal to 2", constraintViolations.iterator().next().getMessage() );
```
Comment (gunnarmorling) [inline osgi/integrationtest/src/test/java/org/hibernate/validator/osgi/integrationtest/OsgiIntegrationTest.java]: Where is that coming from?

**Comment by gunnarmorling** [comment]
@gsmet looks good, added some minor final comments.

File: engine/src/main/java/org/hibernate/validator/internal/util/privilegedactions/SetClassLoader.java (hunk 3 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/util/privilegedactions/SetClassLoader.java]: Yup, I just made it a symmetry of GetClassLoader. I'll change it.

**Comment by gunnarmorling** [comment]
I think that it more was a requirement to use EL at all, it wouldn't have
worked with the default interpolator. I remember that we added that test
now :) (see HV-1039).

File: engine/src/main/java/org/hibernate/validator/internal/engine/messageinterpolation/el/SimpleELContext.java [hunk 5]
```diff
@@ -37,7 +39,9 @@
 	private final VariableMapper variableMapper;
 	private final ELResolver resolver;
 
-	public SimpleELContext() {
+	public SimpleELContext(ExpressionFactory expressionFactory) {
+		putContext( ExpressionFactory.class, expressionFactory );
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/messageinterpolation/el/SimpleELContext.java]: Done.

File: engine/src/main/java/org/hibernate/validator/internal/util/privilegedactions/SetClassLoader.java (hunk 3 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/util/privilegedactions/SetClassLoader.java]: Done.

**Comment by gsmet** [comment]
I'll work on trying to put together a test for that.

**Comment by gsmet** [comment]
@gunnarmorling I tried to write a test using our `IgnoringPackageClassLoader` but I didn't succeed. I think the only way to get it working would be to isolate the test in its own JVM.

So for now, I think I would let it like that. It exercises the fact that we can create a specific ExpressionFactory but we can't really test that it is used.

**Comment by gunnarmorling** [comment]
@gsmet can you update it with my change from #583? I think then we are good to go. Thanks a ton for fighting through this craziness!

**Pull request merged**